tests/integration/test_coredns.py

```diff
@@ -0,0 +1,59 @@
+#!/usr/bin/env python3
+#
+# Copyright 2023 Canonical, Ltd.
+#
+
+import logging
+
+import config
+import pytest
+from conftest import microk8s_kubernetes_cloud_and_model
+from juju.application import Application
+from pytest_operator.plugin import OpsTest
+
+LOG = logging.getLogger(__name__)
+
+
+@pytest.mark.abort_on_fail
+async def test_core_dns(e: OpsTest):
+    # deploy microk8s
+    if "microk8s" not in e.model.applications:
+        await e.model.deploy(
+            config.MK8S_CHARM,
+            application_name="microk8s",
+            config={"hostpath_storage": "true"},
+            channel=config.MK8S_CHARM_CHANNEL,
+            constraints=config.MK8S_CONSTRAINTS,
+        )
+        await e.model.wait_for_idle(["microk8s"])
+
+    # bootstrap a juju cloud on the deployed microk8s
+    async with microk8s_kubernetes_cloud_and_model(e, "microk8s") as (k8s_model, model_name):
+        with e.model_context(k8s_model):
+            LOG.info("Deploy CoreDNS")
+            await e.model.deploy(
+                config.MK8S_CORE_DNS_CHARM,
+                application_name="coredns",
+                channel=config.MK8S_CORE_DNS_CHANNEL,
+                trust=True,
+            )
+            await e.model.wait_for_idle(["coredns"])
+
+            LOG.info("Create offer for dns-provider")
+            await e.model.create_offer("coredns:dns-provider", "coredns")
+
+        try:
+            LOG.info("Consume offer for dns-provider")
+            await e.model.consume(f"admin/{model_name}.coredns", "coredns")
+            LOG.info("Add relation between microk8s and coredns")
+            await e.model.add_relation("microk8s", "coredns")
+            LOG.info("Wait for idle")
+            await e.model.wait_for_idle(["microk8s"])
+
+            with e.model_context(k8s_model):
+                await e.model.wait_for_idle(["coredns"])
+        finally:
+            app: Application = e.model.applications["microk8s"]
+            LOG.info("Remove relation between microk8s and coredns")
+            await app.remove_relation("microk8s:dns", "coredns:dns-provider")
```

Comment:
I'm not sure the juju 3.1 cleaned up this model appropriately

```
ERROR    pytest_operator.plugin:plugin.py:875 ['cannot destroy application "coredns": application is used by 1 consumer']
Traceback (most recent call last):
  File "/home/runner/work/charm-microk8s/charm-microk8s/.tox/integration-3.1/lib/python3.10/site-packages/pytest_operator/plugin.py", line 870, in _destroy
    await entity.destroy(**kwargs)
  File "/home/runner/work/charm-microk8s/charm-microk8s/.tox/integration-3.1/lib/python3.10/site-packages/juju/application.py", line 249, in destroy
    res = await app_facade.DestroyApplication(applications=[client.DestroyApplicationParams(
  File "/home/runner/work/charm-microk8s/charm-microk8s/.tox/integration-3.1/lib/python3.10/site-packages/juju/client/facade.py", line 483, in wrapper
    reply = await f(*args, **kwargs)
  File "/home/runner/work/charm-microk8s/charm-microk8s/.tox/integration-3.1/lib/python3.10/site-packages/juju/client/_client17.py", line 1074, in DestroyApplication
    reply = await self.rpc(msg)
  File "/home/runner/work/charm-microk8s/charm-microk8s/.tox/integration-3.1/lib/python3.10/site-packages/juju/client/facade.py", line 656, in rpc
    result = await self.connection.rpc(msg, encoder=TypeEncoder)
  File "/home/runner/work/charm-microk8s/charm-microk8s/.tox/integration-3.1/lib/python3.10/site-packages/juju/client/connection.py", line 661, in rpc
    raise errors.JujuError(err_results)
juju.errors.JujuError: ['cannot destroy application "coredns": application is used by 1 consumer']
```


```suggestion
            await app.remove_relation("dns", "coredns")
```

Reply:
Hey @addyess, this is something that keeps coming. I've tried locally by removing the relation (and the correct names) and even without removing relations. But during the resetting, this pops up.

Reply:
I've raised https://github.com/canonical/charm-microk8s/pull/100 into this branch which seems to have removed the errors in the tests i spotted a few days ago